src/AspNet.Security.OAuth.Alipay/AlipayAuthenticationHandler.cs

```diff
@@ -130,40 +144,158 @@ protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OA
+                            /* Status: */ statusCode,
+                            /* Message: */ payload.RootElement.GetString("msg"));
 
                 throw new HttpRequestException("An error occurred while retrieving user information.");
             }
 
-            identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, alipay_user_info_share_response.RootElement.GetString("user_id"), ClaimValueTypes.String, Options.ClaimsIssuer));
+            identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, payload.RootElement.GetString("user_id"), ClaimValueTypes.String, Options.ClaimsIssuer));
 
             var principal = new ClaimsPrincipal(identity);
-            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, alipay_user_info_share_response.RootElement);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+
             context.RunClaimActions();
 
             await Options.Events.CreatingTicket(context);
             return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
         }
 
-        protected override string FormatScope() => string.Join(",", Options.Scope);
-        private IDictionary<string, string> RSA2Sign(SortedDictionary<string, string> source)
+        /// <summary>
+        /// get payload as an asynchronous operation.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        /// <param name="nodeName">Name of the node.</param>
+        /// <returns>Task&lt;JsonDocument&gt;.</returns>
+        private async Task<JsonDocument> getPayloadAsync(HttpResponseMessage message, string nodeName)
         {
-            var textparams = source.Where(p => !string.IsNullOrEmpty(p.Value)).Select(p => p.Key + "=" + p.Value);
-            if (!textparams.Any())
-                throw new HttpRequestException("sign fail");
-            var rsa2 = new RSAHelper(RSAType.RSA2, Encoding.UTF8, Options.ClientSecret, Options.AlipayPublicKey);
-            source.Add("sign", rsa2.Sign(string.Join("&", textparams)));
-            return source;
+            var messageContent = await message.Content.ReadAsStringAsync();
+
+            //Logger.LogInformation(messageContent);
+
+            var rootJson = JsonDocument.Parse(messageContent);
+            var payload = JsonDocument.Parse(rootJson.RootElement.GetString(nodeName));
+
+            return payload;
+        }
+
+        /// <summary>
+        /// Gets the RSA2 signature.
+        /// </summary>
+        /// <param name="source">The source.</param>
+        /// <param name="privateKey">The private key.</param>
+        /// <returns>System.String.</returns>
+        private string getRSA2Signature(SortedDictionary<string, string> source, [NotNull] string privateKey)
+        {
+            var textParams = source
```

Comment:
I still think it would be better to build this up using `QueryHelpers.AddQueryString()` to ensure it's correctly encoded.

Reply:
Sorry, i don't know how to modify?

Reply:
url encode and signature has not related.

Reply:
This is what I was thinking:

```csharp
string dataToSign = string.Empty;

foreach (var pair in source)
{
    if (string.Equals(pair.Key, "sign", StringComparison.Ordinal))
    {
        continue;
    }

    dataToSign = QueryHelpers.AddQueryString(dataToSign, pair.Key, pair.Value);
}


byte[] dataBytes = Encoding.UTF8.GetBytes(dataToSign[1..]);
```

Reply:
1, sort by ascii.
2, except empty value and "sign".
3, signature (must original value).
4, url encode.
5, send.

Reply:
How about this instead? It saves allocating lots of intermediate strings, and also from what I can see the sorted dictionary never contains `"sign"` anymore, so the code to filter that isn't needed:

```csharp
private string GetRSA2Signature(SortedDictionary<string, string> source)
{
    var builder = new StringBuilder();

    foreach (var pair in source)
    {
        if (string.IsNullOrEmpty(pair.Value))
        {
            continue;
        }

        builder.Append(pair.Key)
               .Append('=')
               .Append(pair.Value)
               .Append('&');
    }

    builder.Remove(builder.Length - 1, 1);

    byte[] dataBytes = Encoding.UTF8.GetBytes(builder.ToString());

    using var privateKeyRsaProvider = CreateAlgorithm();
    byte[] signatureBytes = privateKeyRsaProvider.SignData(dataBytes, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);

    return Convert.ToBase64String(signatureBytes);
}
```